### ISSUE TITLE
Issues with the data reader fetching yahoo finance

### ISSUE THREAD

**Issue by Crowbeezy** [pydata/pandas-datareader#315]
Apologies first issue/comment on GitHub.  I will review proper protocol.  Please correct me if this is not the correct place to put this. 

---------------------------------------------------------------------------
RemoteDataError                           Traceback (most recent call last)
<ipython-input-17-4b8330445ce8> in <module>()
      4 end = dt.datetime(2017, 5, 8)
      5 
----> 6 INPX = data.DataReader(INPX ,'yahoo', start, end)
      7 
      8 #Convert Volume from Int to Float

C:\Users\randomname\AppData\Local\Continuum\Anaconda3\lib\site-packages\pandas_datareader\data.py in DataReader(name, data_source, start, end, retry_count, pause, session)
     92                                 adjust_price=False, chunksize=25,
     93                                 retry_count=retry_count, pause=pause,
---> 94                                 session=session).read()
     95 
     96     elif data_source == "yahoo-actions":

C:\Users\randomname\AppData\Local\Continuum\Anaconda3\lib\site-packages\pandas_datareader\yahoo\daily.py in read(self)
     75     def read(self):
     76         """ read one data from specified URL """
---> 77         df = super(YahooDailyReader, self).read()
     78         if self.ret_index:
     79             df['Ret_Index'] = _calc_return_index(df['Adj Close'])

C:\Users\randomname\AppData\Local\Continuum\Anaconda3\lib\site-packages\pandas_datareader\base.py in read(self)
    176             df = self._dl_mult_symbols(self.symbols.index)
    177         else:
--> 178             df = self._dl_mult_symbols(self.symbols)
    179         return df
    180 

C:\Users\randomname\AppData\Local\Continuum\Anaconda3\lib\site-packages\pandas_datareader\base.py in _dl_mult_symbols(self, symbols)
    195         if len(passed) == 0:
    196             msg = "No data fetched using {0!r}"
--> 197             raise RemoteDataError(msg.format(self.__class__.__name__))
    198         try:
    199             if len(stocks) > 0 and len(failed) > 0 and len(passed) > 0:

RemoteDataError: No data fetched using 'YahooDailyReader'

**Comment by rgkimball** [comment]
Can you provide a sample that replicates your issue? This works for me:

```
start = datetime(2016, 12, 31)
end = datetime.now()
INPX = data.DataReader('INPX', 'yahoo', start, end)
```

**Comment by Crowbeezy** [comment]
On Thu, May 11, 2017 at 9:49 PM, Rob Kimball <notifications@github.com>
wrote:

> start = datetime(2016, 12, 31) end = datetime.now() INPX =
> data.DataReader('INPX', 'yahoo', start, end)



I'm not sure what the problem was. Perhaps using 'end = datetime.today()'?

You're code worked.  Thank you for your help!

**Comment by benpillet** [comment]
From my requirements.txt:
```
pandas-datareader==0.4.0
pandas==0.20.1
```
and in python shell:
```
from datetime import *
import pandas_datareader.data as data
start = datetime(2016, 12, 31)
end = datetime.now()
INPX = data.DataReader('INPX', 'yahoo', start, end)
```
with error:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "./venv3.5/lib/python3.5/site-packages/pandas_datareader/data.py", line 117, in DataReader
    session=session).read()
  File "./venv3.5/lib/python3.5/site-packages/pandas_datareader/yahoo/daily.py", line 77, in read
    df = super(YahooDailyReader, self).read()
  File "./venv3.5/lib/python3.5/site-packages/pandas_datareader/base.py", line 157, in read
    params=self._get_params(self.symbols))
  File "./venv3.5/lib/python3.5/site-packages/pandas_datareader/base.py", line 74, in _read_one_data
    out = self._read_url_as_StringIO(url, params=params)
  File "./venv3.5/lib/python3.5/site-packages/pandas_datareader/base.py", line 85, in _read_url_as_StringIO
    response = self._get_response(url, params=params)
  File "./venv3.5/lib/python3.5/site-packages/pandas_datareader/base.py", line 120, in _get_response
    raise RemoteDataError('Unable to read URL: {0}'.format(url))
pandas_datareader._utils.RemoteDataError: Unable to read URL: http://ichart.finance.yahoo.com/table.csv?f=2017&ignore=.csv&b=31&c=2016&g=d&a=11&d=4&s=INPX&e=16
```

I have a feeling yahoo updated their endpoint to be something else.  I get a 502 when I try curl too.  The link at `https://finance.yahoo.com/quote/SPY/history?p=SPY` points to `https://query1.finance.yahoo.com/v7/finance/download/SPY?period1=1492372898&period2=1494964898&interval=1d&events=history&crumb=MLOX17FWABw`

**Comment by benpillet** [comment]
Looks like there's also a cookie that needs to be sent in order to avoid a `401 Unauthorized`.  [https://www.elitetrader.com/et/threads/yahoo-historical-data-did-they-change-the-url-recently.309554/](url)

**Comment by rgkimball** [comment]
Not sure if the icharts failure is a temporary problem, but I submitted a WIP PR (above) to replace the request structure. Even if icharts does come back online, may be a good idea to implement a backup.

**Comment by IvanTrendafilov** [comment]
I don't have the time to fix this in the library, but, essentially, there is another API endpoint that one can use. It's `https://query1.finance.yahoo.com`. But it requires a matching cookie and breadcrumb.  I wrote a little PhantomJS script to get it, whilst I was working on it: https://github.com/IvanTrendafilov/YahooFinanceAPITokens

You can also get it from Chrome dev tools in the network tab.

**Comment by IvanTrendafilov** [comment]
They've confirmed icharts isn't coming back, so @rgkimball's patch should certainly go in.

https://forums.yahoo.net/t5/Yahoo-Finance-help/Is-Yahoo-Finance-API-broken/td-p/250503/page/3

**Comment by bkcollection** [comment]
@rgkimball @IvanTrendafilov , can the fix be pip install upgrade for the ease for beginner?

**Comment by Franlodo** [comment]
Yahoo has change the URL, and the way the use date. Now date are Unixtime.
For example to get historic cvs from AAPL :
https://query1.finance.yahoo.com/v7/finance/download/AAPL?period1=1492510098&period2=1495102098&interval=1d&events=history&crumb=ydacXMYhzrn

period1 or period2 is date in (unixtime = (Human time - 25568) * 86400) but you must check your timezone, for example my place is Europe, then I have UTC+2 and I must sustract 7200 seconds. So my formula is (Human time - 25568) * 86400) -7200; where Human time is the time (d/mm/yyyy), 25568 is the number of days  from 01/01/1900 till 01/01/1970 (This is because i do it in Excel and this is the minimun date), 86400 are the seconds in a day and 7200 are the number of seconds in my 2 hours difference with UTC

Interval is day, week or month
Events history is historic data prices, div|split&filter=split, for splits and div|split&filter=div for dividens
Crumb is the cookie ... I don't really know how it works, but I have with the same since monday.

I'm using this for update my data in Excel and it works and now I don't need to wait until morning to get the historical data because it's available at less an hour after the market close (I´m talking about american markets)

I apologyze for not to be fluent in english.

I hope this help

**Comment by bkcollection** [comment]
@Franlodo DO you try do download suing the new link for like 1000 stocks, will it blocked? The old API seems has no limitation but I am curious if the new one still allow that. Hope you can try on it to validate

**Comment by Franlodo** [comment]
The link is to get the csv file in the web, it must run for 2 or for 2000; In my Excel file I have nearly 200 and run properly.

You can get 1000 of csv files and import from pandas, it will be the same made it saving files or "in the air"
I had post here because the error reported for pandas-datareader was the url

Anyway, I will try and comment.

**Comment by rgkimball** [comment]
@bkcollection This will be available once the bugs are ironed out and the pull request is merged into the main repository.

**Comment by bkcollection** [comment]
@rgkimball how optimistic the bugs can be fixed?

**Comment by rgkimball** [comment]
@bkcollection The latest commits of #331 about wraps it up. It's a little frustrating that the new API drops out periodically, but you can now pull any historical price range, splits and dividends. I haven't found a new interface for Yahoo Options - this may be permanently removed, but I'm happy to implement if someone finds the endpoint. Same thing for index constituents.

All of the failing tests on my PR are due to Eurostat or Yahoo's API sporadically failing. Notice that there is inconsistency across different tests for which ones are failing - appears to be random. Hoping we have some other people find time pull down the code and give it a thorough review before the maintainers make a decision on merging it in.

**Comment by bkcollection** [comment]
How to get the implement the fix into my current module? How is the drop
out happen? Can you explain more?

**Comment by gusutabopb** [comment]
@bkcollection:

For a temporary fix (until this [PR](https://github.com/pydata/pandas-datareader/pull/331) gets merged), try:

```bash
$ git clone https://github.com/rgkimball/pandas-datareader
$ cd pandas-datareader
$ git checkout fix-yahoo
$ pip install -e .
```

On Python:
```python
import pandas_datareader as pdr
print(pdr.__version__)  # Make sure it is '0.4.1'.
```

I originally wrote this as an answer to [this](http://stackoverflow.com/questions/44087113/error-with-pulling-data-from-yahoo-finance) Stackoverflow question

**Comment by hmz123** [comment]
@gusutabopb
 The latest version is "0.4.0" not it?

**Comment by gusutabopb** [comment]
@hmz123 
The latest on `master`/pypi is 0.4.0. The [PR](https://github.com/pydata/pandas-datareader/pull/331) proposed by @rgkimball makes it 0.4.1.
The above instructions are for those who do not want to / can't wait for an official upgrade. See the commits of the `fix-yahoo` patch branch here: https://github.com/rgkimball/pandas-datareader/commits/fix-yahoo

**Comment by bkcollection** [comment]
seems like still have some errors. Not sure when will be a fixed release 0.4.1. Hopefully in this week.

**Comment by arose13** [comment]
@bkcollection what do you mean by some errors?
Do you mean being blocked by yahoo if you request a handful of symbols?

**Comment by bkcollection** [comment]
@arose13 you have blocked for using too many symbols? How many is it?

**Comment by arose13** [comment]
@bkcollection 4-6 in 10 mins got blocked for ~12h

**Comment by bkcollection** [comment]
@arose13 That is terrible. I would like to download like 1000 tickers. 4 to 6 tickers and get blocked is by well not working to me

**Comment by Franlodo** [comment]
Now I'm downloading 6750 tickers ... all the list in Nasdaq website; in monthly, weekly and daily

**Comment by bkcollection** [comment]
@Franlodo Thanks.  Looks like it works as previously 
@rgkimball Is there any schedule that your fix will be in 0.4.1 soon?

**Comment by cpankaj** [comment]
I was able to fetch data from Yahoo Finance and implemented the fix for ruby
https://github.com/cpankaj/yquotes

Solution for 401/unauthorized is to get fresh cookie and crumb and retry

**Comment by bkcollection** [comment]
@rgkimball any schedule that your fix for yahoo will be in 0.4.1?

**Comment by hmz123** [comment]
Hi,
Someone can test this,i don't have my computer with me
https://github.com/ranaroussi/fix-yahoo-finance

**Comment by bkcollection** [comment]
@hmz123  python 2.7 not supported. 
Seems like have to wait 0.4.1 which have no schedule yet.

**Comment by deios0** [comment]
Unfortunately, historical data from http://ichart.finance.yahoo.com/table.csv is no longer working. But the thing with finding and keeping crumbs is not obvious and also since it's non-official, I guess it could be changed too. That's why I tried to find free or at least cheap alternative and found this one: https://eodhistoricaldata.com/.

It provides the data in plain CSV, exactly the same format as Yahoo Finance, then it's easy to switch, just by replacing URLs.

**Comment by femtotrader** [comment]
EOD Historical Data wrote a blog post
https://eodhistoricaldata.com/knowledgebase/adapt-old-yahoo-scripts-eod-historical-data/
but it's only available for free with AAPL.US symbol... for others symbol it's a paid API

**Comment by ehtom** [comment]
Has anyone else noticed that the data quality is significantly worse than before? There seem to be a number of large gaps (filled with 'null') for several tickers. Data is flat out wrong in some cases (e.g. 20% daily moves in low, for example)...

This seems to also be a problem for the new yahoo finance website (look at BA.L, for example).

Does anyone know of any other high quality free sources of EOD data, or of reliable paid-for data?

**Comment by femtotrader** [comment]
Here is a short script for downloading data from EOD historical data

import pandas as pd
from pandas.compat import StringIO
from pandas_datareader._utils import _init_session

```python
def get_eod_data(symbol="AAPL.US", api_token="OeAFFmMliFG5orCUuwAKQ8l4WWFQ67YX", session=None):
    session = _init_session(session)
    url = "https://eodhistoricaldata.com/api/eod/%s" % symbol
    params = {
        "api_token": api_token
    }
    r = session.get(url, params=params)
    df = pd.read_csv(StringIO(r.text), skipfooter=1, parse_dates=[0], index_col=0)
    return df

df = get_eod_data()
```

be aware that it's only available for free with AAPL.US symbol... if you want more you will have to pay ;-(

**Comment by jotbe** [comment]
What about [Alphavantage](http://www.alphavantage.co/) as a data provider? I tinkered a bit with their API, it is a bit slow sometimes and I didn't check the quality of their data yet. They offer plenty of indicators and even (near-)realtime data as JSON.

**Comment by deios0** [comment]
@jotbe I tried them 1-2 years ago, quality is not good and data not updated regular.

**Comment by femtotrader** [comment]
Here is some code to have a look (only intraday, daily, weekly, monthly data)

```python
import requests
import pandas as pd
from pandas.io.common import urlencode
from pandas.compat import StringIO
import json
from pandas_datareader._utils import _init_session
from pandas.tseries.frequencies import to_offset

def get_data_alphavantage(symbol, interval=None, outputsize=None, api_key=None, adjusted=False, session=None):
    session = _init_session(session)
    url = "http://www.alphavantage.co/query"
    
    # apikey
    if api_key is None:
        api_key = "demo"

    # function
    d_functions = {
        to_offset("D").freqstr: "TIME_SERIES_DAILY",
        to_offset("W").freqstr: "TIME_SERIES_WEEKLY",
        to_offset("M").freqstr: "TIME_SERIES_MONTHLY",
    }
    try:
        if adjusted and to_offset(interval).freqstr == "D":
            function_api = "TIME_SERIES_DAILY_ADJUSTED"
        else:            
            function_api = d_functions[to_offset(interval).freqstr]
    except KeyError:
        function_api = "TIME_SERIES_INTRADAY"
    
    # interval
    if interval is None:
        interval = "D"
    d_intervals = {
        to_offset("1T").freqstr: "1min",
        to_offset("5T").freqstr: "5min",
        to_offset("15T").freqstr: "15min",
        to_offset("30T").freqstr: "30min",
        to_offset("H").freqstr: "60min",
        to_offset("D").freqstr: "daily",
        to_offset("W").freqstr: "weekly",
        to_offset("M").freqstr: "monthly"
    }
    try:
        interval_api = d_intervals[to_offset(interval).freqstr]
    except KeyError:
        interval_api = "60min"

    # outputsize
    if outputsize is None:
        outputsize = "compact"

    params = {
        "function": function_api,
        "symbol": symbol,
        "interval": interval_api,
        "outputsize": outputsize,
        "apikey": api_key
    }
    r = session.get(url, params=params)
    if params is not None and len(params) > 0:
        url_long = url + "?" + urlencode(params)
    else:
        url_long = url
    print(url_long)
    if r.status_code == requests.codes.ok:
        dat = r.json()
        key_dat = list(dat.keys())[1]  # ugly
        ts = dat[key_dat]
        df = pd.DataFrame(ts).T
        df = df.rename(columns={
            "1. open": "Open",
            "2. high": "High",
            "3. low": "Low",
            "4. close": "Close",
            "5. volume": "Volume",
        })
        for col in ["Open", "High", "Low", "Close"]:
            if col in df.columns:
                df[col] = df[col].astype(float)
        df["Volume"] = df["Volume"].astype(int)
        df.index = pd.to_datetime(df.index)
        df.index.name = "Date"
        return df
    else:
        params["apikey"] = "HIDDEN"
        raise Exception(r.status_code, r.reason, url_long)

import pandas as pd
pd.set_option("max_rows", 10)
import datetime        
import requests_cache

expire_after = datetime.timedelta(days=1)
session = requests_cache.CachedSession(cache_name='cache', backend='sqlite', expire_after=expire_after)

api_key = "YOURAPIKEY"
df = get_data_alphavantage("MSFT", interval="W", api_key=api_key, session=session)
print(df)
print(df.dtypes)


df = get_data_alphavantage("MSFT", interval="15Min", api_key=api_key, session=session)
print(df)
print(df.dtypes)
```

**Comment by justinlent** [comment]
+1 for this as a reasonable option/implementation

** Sent from iPhone **
________________________________

**Comment by deios0** [comment]
@justinlent why not to use https://eodhistoricaldata.com/. They have nice data for cheap prices. Do you have any concerns here?

**Comment by jainraje** [comment]
FYI, 
I'm using https://github.com/ranaroussi/fix-yahoo-finance. the historical data download appears to work quite well. 

To my surprise (or maybe not so much these days) Yahoo Finance did not return any historical data for certain key US tickers. For example: CAT, COP, GRMN and many others. Posting this mostly as FYI and to see if any others *are* able to download CAT, COP or GRMN historic data. 

Perhaps yahoo finance engineers broke a few things (data quality, availability of certain tickers, other?) as they migrated to the new API?

**Comment by justinlent** [comment]
Agreed — that EOD data works very well too. Was more meaning as something
that’s **free**, the alphavantage looks decent and is easy and
straightforward for someone looking for data to play around with and
doesn’t want to commit to a constant monthly subscription

**Comment by bkcollection** [comment]
What country akpha vantage support?

**Comment by deios0** [comment]
@jotbe "They offer plenty of indicators and even (near-)realtime data as JSON. I didn't find out yet, who is offering this service." Btw, you are really need technical indicators with API? How do you use them?

**Comment by krinkere** [comment]
At this point instead of waiting or trying to fix yahoo, i switched to quandl... at least for my purposes it does the job.

**Comment by adminho** [comment]
I followed https://github.com/ranaroussi/fix-yahoo-finance. It's work

**Comment by deios0** [comment]
Yes, but they write there: **"a temporary fix**". I think soon you'll get the same problems with Yahoo.

**Comment by javadba** [comment]
fyi: the fix from ranaraoussi requires python 3.4+

**Comment by liuyigh** [comment]
@javadba a python 2.7-compatible pull request was merged 5 days ago. FYI.

**Comment by aisthesis** [comment]
rgkimball's version 0.4.1 is still working for me. Is there a reason why it isn't being merged?

**Comment by jreback** [comment]
it needs to pass tests and respond to comments

**Comment by rgkimball** [comment]
I believe comments have been covered - I will try to work on tests next week. Did I miss anything?

Rob Kimball

**Comment by jainraje** [comment]
@rgkimball if you can provide me instructions on how to install your version, i will run some tests and provide feedback and results. i'm currently using the fix-yahoo_finance developed by ranaroussi.

**Comment by Harrymon12** [comment]
Yahoo finance API   is not available anymore. I have moved to MarketXLS after this change, much   more reliable data.

**Comment by m3nu** [comment]
@Harrymon12 is a paid troll that spams about "MarketXLS" all over the internet. Here some examples of his "work":

- http://www.databison.com/stock-chart-with-scroll-and-zoom/
- https://www.quantnet.com/threads/historical-stock-quotes-in-excel.8060/
- https://davidwalsh.name/stock-quotes-command-line

**Comment by Harrymon12** [comment]
@m3nu  Ahw. Oh really? I am just giving the facts about MarketXLS.
What is the problem about that?
You can check their forums if it really was a spam.
How can you say it is spam when you did not use MarketXLS yet?
Are you even normal or just a poser?

**Comment by javadba** [comment]
@Harrymon12  This is a pandas site. If you have some tips about helping out on PANDAS and specifically this issue please feel free to do so. Otherwise your posts ARE spam.

**Comment by Harrymon12** [comment]
@javadba  I understand. Thank you. :)

**Comment by JECSand** [comment]
@javadba 
@justinlent 
@rgkimball 

My replacement for the old yahoo-finance module, **YahooFinancials**, can get all of the historical price data needed by pandas' users. **YF can return daily, weekly, and monthly historical price and volume JSON data for all stocks, ETFs, indices, cryptocurrencies, currencies, and commodity futures available on Yahoo Finance.** Most Stackoverflow questions I have encountered regarding the module seem to revolve around getting it to work with Pandas. As long as Yahoo Finance is running on it's new React setup (why they killed the old API late last year, they got a new web app), my module will get the financial data.

Usage Example:
```
from yahoofinancials import YahooFinancials

yahoo_financials = YahooFinancials('WFC')
print(yahoo_financials.get_historical_price_data("2018-07-10", "2018-08-10", "monthly"))
```
Returns
```
{
    "WFC": {
        "currency": "USD",
        "eventsData": {
            "dividends": {
                "2018-08-01": {
                    "amount": 0.43,
                    "date": 1533821400,
                    "formatted_date": "2018-08-09"
                }
            }
        },
        "firstTradeDate": {
            "date": 76233600,
            "formatted_date": "1972-06-01"
        },
        "instrumentType": "EQUITY",
        "prices": [
            {
                "adjclose": 57.19147872924805,
                "close": 57.61000061035156,
                "date": 1533096000,
                "formatted_date": "2018-08-01",
                "high": 59.5,
                "low": 57.08000183105469,
                "open": 57.959999084472656,
                "volume": 138922900
            }
        ],
        "timeZone": {
            "gmtOffset": -14400
        }
    }
}
```

**Anyway I'd be happy to fork a branch and build the price data from YahooFinancials into the panda-datareader's get_data_yahoo() method if you all want. I'd also be happy to work with one of your contributors to do so as well.** Just let me know and I'd be happy to help!

More details at:
https://github.com/JECSand/yahoofinancials

**Comment by magicmathmandarin** [comment]
Hi, Yahoo Finance is working for me.  I am confused as to why you guys saying it is deprecated?

**Comment by GrechTsangWL** [comment]
Maybe you are out of the great wall. Maybe others.

发件人: Sarah Chen [mailto:notifications@github.com]
发送时间: 2019年4月2日 6:44
收件人: pydata/pandas-datareader <pandas-datareader@noreply.github.com>
抄送: Subscribed <subscribed@noreply.github.com>
主题: Re: [pydata/pandas-datareader] Issues with the data reader fetching yahoo finance (#315)


Hi, Yahoo Finance is working for me. I am confused as to why you guys saying it is deprecated?

**Comment by bashtage** [comment]
It was deprecated, and then it was restored using a new implementation.

**Comment by magicmathmandarin** [comment]
Then it would be good to stay in the official  documentation for version
0.70.  Otherwise people get scared of using it or confused:)

**Comment by abdoulayegk** [comment]
Can anyone help me fix this error

NotImplementedError: data_source=datetime.datetime(2015, 1, 1, 0, 0) is not implemented
Traceback:
File "/home/balde/.local/lib/python3.8/site-packages/streamlit/ScriptRunner.py", line 322, in _run_script
    exec(code, module.__dict__)
File "/home/balde/Desktop/Projects/StockProject/stock_app.py", line 14, in <module>
    globals()[stock] = DataReader(stock,start, end)
File "/home/balde/.local/lib/python3.8/site-packages/pandas/util/_decorators.py", line 214, in wrapper
    return func(*args, **kwargs)
File "/home/balde/.local/lib/python3.8/site-packages/pandas_datareader/data.py", line 376, in DataReader
    raise NotImplementedError(msg)

**Comment by krb1971** [comment]
I recently started to learn python for finance. From this thread, I understood at some level the yahoo fin package related issues but while I started to run the following code, 'ticker' not found error is persistent. Could you please check it and guide me further? 

Here are the libraries used and the code: 

import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout, Bidirectional
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from yahoo_fin import stock_info as si
from collections import deque

import numpy as np
import pandas as pd
import random

import matplotlib
import matplotlib.pyplot as plt
import matplotlib.ticker as mticker
import matplotlib.dates as mdates
import numpy as np

def load_data(ticker, n_steps=50, scale=True, shuffle=True, lookup_step=1, 
                test_size=0.2, feature_columns=['adjclose', 'volume', 'open', 'high', 'low']):
    # see if ticker is already a loaded stock from yahoo finance
    if isinstance(ticker, str):
        # load it from yahoo_fin library
        df = si.get_data(ticker)
    elif isinstance(ticker, pd.DataFrame):
        # already loaded, use it directly
        df = ticker
    # this will contain all the elements we want to return from this function
    result = {}
    # we will also return the original dataframe itself
    result['df'] = df.copy()
    # make sure that the passed feature_columns exist in the dataframe
    for col in feature_columns:
        assert col in df.columns, f"'{col}' does not exist in the dataframe."
    if scale:
        column_scaler = {}
        # scale the data (prices) from 0 to 1
        for column in feature_columns:
            scaler = preprocessing.MinMaxScaler()
            df[column] = scaler.fit_transform(np.expand_dims(df[column].values, axis=1))
            column_scaler[column] = scaler

        # add the MinMaxScaler instances to the result returned
        result["column_scaler"] = column_scaler
    # add the target column (label) by shifting by `lookup_step`
    df['future'] = df['adjclose'].shift(-lookup_step)
    # last `lookup_step` columns contains NaN in future column
    # get them before droping NaNs
    last_sequence = np.array(df[feature_columns].tail(lookup_step))
    # drop NaNs
    df.dropna(inplace=True)
    sequence_data = []
    sequences = deque(maxlen=n_steps)
    for entry, target in zip(df[feature_columns].values, df['future'].values):
        sequences.append(entry)
        if len(sequences) == n_steps:
            sequence_data.append([np.array(sequences), target])
    # get the last sequence by appending the last `n_step` sequence with `lookup_step` sequence
    # for instance, if n_steps=50 and lookup_step=10, last_sequence should be of 59 (that is 50+10-1) length
    # this last_sequence will be used to predict in future dates that are not available in the dataset
    last_sequence = list(sequences) + list(last_sequence)
    # shift the last sequence by -1
    last_sequence = np.array(pd.DataFrame(last_sequence).shift(-1).dropna())
    # add to result
    result['last_sequence'] = last_sequence
    # construct the X's and y's
    X, y = [], []
    for seq, target in sequence_data:
        X.append(seq)
        y.append(target)
    # convert to numpy arrays
    X = np.array(X)
    y = np.array(y)
    # reshape X to fit the neural network
    X = X.reshape((X.shape[0], X.shape[2], X.shape[1]))
    # split the dataset
    result["X_train"], result["X_test"], result["y_train"], result["y_test"] = train_test_split(X, y, test_size=test_size, shuffle=shuffle)
    # return the result
    return result

# load the data
data = load_data(ticker, N_STEPS, lookup_step=LOOKUP_STEP, test_size=TEST_SIZE, feature_columns=FEATURE_COLUMNS)


THIS GIVES FOLLOWING ERROR:

NameError                                 Traceback (most recent call last)
<ipython-input-15-105f1f13d7b6> in <module>()
      1 # load the data
----> 2 data = load_data(ticker, N_STEPS, lookup_step=LOOKUP_STEP, test_size=TEST_SIZE, feature_columns=FEATURE_COLUMNS)
      3 
      4 # save the dataframe
      5 data["df"].to_csv(ticker_data_filename)

NameError: name 'ticker' is not defined

**Comment by rgkimball** [comment]
> # load the data
> data = load_data(ticker, N_STEPS, lookup_step=LOOKUP_STEP, test_size=TEST_SIZE, feature_columns=FEATURE_COLUMNS)
> 
> THIS GIVES FOLLOWING ERROR:
> 
> NameError Traceback (most recent call last)
> in ()
> 1 # load the data
> ----> 2 data = load_data(ticker, N_STEPS, lookup_step=LOOKUP_STEP, test_size=TEST_SIZE, feature_columns=FEATURE_COLUMNS)
> 3
> 4 # save the dataframe
> 5 data["df"].to_csv(ticker_data_filename)
> 
> NameError: name 'ticker' is not defined

@krb1971,

The error you've shared doesn't have anything to do with the pandas datareader package. You need to define ticker before passing it into the function.

**Comment by krb1971** [comment]
Thank you for responding. When I defined load_data as following, it did not ask for defining 'ticker'

def load_data(ticker, n_steps=50, scale=True, shuffle=True, lookup_step=1,
test_size=0.2, feature_columns=['adjclose', 'volume', 'open', 'high', 'low']):
....................

But when I use ticker below, it complains as "ticker" not defined. As I am novice to this, kindly guide.

data = load_data(ticker, N_STEPS, lookup_step=LOOKUP_STEP, test_size=TEST_SIZE, feature_columns=FEATURE_COLUMNS)

**Comment by Franlodo** [comment]
Good morning ... That's not the place for this kind of questions ...

You could look for  correct answers in
https://bit.ly/35c8jlI or in shorturl.at/mMW47

If you add ticket = "whatever ticker stock you want" before call the
función, the first part of your problem will be solved.

Best regards
